fread internal error on invalid file

On MacOS with debug build:
```
>>> import datatable as dt
>>> dt.fread("{\n  \"cells\": [\n    {\n\"import numpy \\n\",\n")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'n_sample_lines <= allocnrow' failed in src/core/csv/reader_fread.cc, line 666
```
On PPC64 it results in
```
datatable.exceptions.MemoryError: Unable to allocate memory of size 18446744073709551612[errno 12] Cannot allocate memory
```

_Originally posted by @pseudotensor at https://h2oai.slack.com/archives/C0BECF81M/p1594140332439500_

--- a/src/core/csv/reader_fread.cc
+++ b/src/core/csv/reader_fread.cc
@@ -640,8 +640,8 @@ void FreadReader::detect_column_types()
     meanLineLen = sumLen;
   } else {
     size_t bytesRead = static_cast<size_t>(eof - sof);
-    meanLineLen = sumLen/n_sample_lines;
-    size_t estnrow = static_cast<size_t>(std::ceil(bytesRead/meanLineLen));  // only used for progress meter and verbose line below
+    meanLineLen = std::max(sumLen/n_sample_lines, 1.0);
+    size_t estnrow = static_cast<size_t>(std::ceil(bytesRead/meanLineLen));
     double sd = std::sqrt( (sumLenSq - (sumLen*sumLen)/n_sample_lines)/(n_sample_lines-1) );
     allocnrow = std::max(static_cast<size_t>(bytesRead / fmax(meanLineLen - 2*sd, minLen)),
                          static_cast<size_t>(1.1*estnrow));

--- a/tests/fread/test_fread_issues.py
+++ b/tests/fread/test_fread_issues.py
@@ -535,3 +535,9 @@ def test_issue2458(capsys):
     mm = re.search(r"\(\s?(\d+)%\) saving into the output frame", out)
     time_percent = int(mm.group(1))
     assert time_percent < 30
+
+
+def test_issue2523():
+    # Make sure no AssertionError occurs here
+    with pytest.raises(IOError):
+        dt.fread("{\n  \"cells\": [\n    {\n\"import numpy \\n\",\n")

